src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.spec.ts

```diff
@@ -1,33 +1,505 @@
+    const navSpy: Spy = spyOn(navigationService, 'navigateWithSuccessMessage');
+    spyOn(feedbackSessionsService, 'moveSessionToRecycleBin').and.returnValue(of(true));
+    component.deleteExistingSessionHandler();
+
+    expect(navSpy.calls.count()).toEqual(1);
+    expect(navSpy.calls.mostRecent().args[1]).toEqual('/web/instructor/sessions');
+    expect(navSpy.calls.mostRecent().args[2]).toEqual('The feedback session has been deleted. '
+      + 'You can restore it from the deleted sessions table below.');
+  });
+
+  it('should create new question', () => {
+    component.isAddingQuestionPanelExpanded = true;
+    component.questionEditFormModels = [testQuestionEditFormModel1];
+    component.feedbackQuestionModels = new Map().set(testFeedbackQuestion1.feedbackQuestionId, testFeedbackQuestion1);
+    spyOn(feedbackQuestionsService, 'createFeedbackQuestion').and.returnValue(of(testFeedbackQuestion2));
+    component.createNewQuestionHandler();
+    expect(component.questionEditFormModels.length).toEqual(2);
+    expect(component.feedbackQuestionModels.has(testFeedbackQuestion2.feedbackQuestionId)).toEqual(true);
+  });
+
+  it('should open modal and copy current session', () => {
+    class MockNgbModalRef {
+      componentInstance: any = {
+        newFeedbackSessionName: '',
+        courseCandidates: [],
+        sessionToCopyCourseId: '',
+      };
+      result: Promise<any> = Promise.resolve({
+        newFeedbackSessionName: 'new feedback session',
+        copyToCourseList: ['testId2'],
+      });
+    }
+    const mockModalRef: MockNgbModalRef = new MockNgbModalRef();
+    const copiedFeedbackSession: FeedbackSession = JSON.parse(JSON.stringify(testFeedbackSession));
+    copiedFeedbackSession.courseId = 'testId2';
+
+    component.feedbackSessionName = testFeedbackSession.feedbackSessionName;
+    component.courseId = testCourse1.courseId;
+    spyOn(courseService, 'getInstructorCoursesThatAreActive').and.returnValue(of({ courses: [testCourse2] }));
+    spyOn(feedbackSessionsService, 'getFeedbackSession').and.returnValue(testFeedbackSession);
+    spyOn(feedbackSessionsService, 'createFeedbackSession').and.returnValue(of(copiedFeedbackSession));
+    spyOn(feedbackQuestionsService, 'getFeedbackQuestions').and.returnValue(of({ questions: [testFeedbackQuestion1] }));
+    spyOn(feedbackQuestionsService, 'createFeedbackQuestion').and.returnValue(of(testFeedbackQuestion1));
+
+    spyOn(ngbModal, 'open').and.returnValue(mockModalRef);
+
+    expect(component.copyCurrentSession()).resolves.toBe(undefined);
+    expect(ngbModal.open).toHaveBeenCalledWith(CopySessionModalComponent);
+    expect(mockModalRef.componentInstance.newFeedbackSessionName).toEqual(testFeedbackSession.feedbackSessionName);
+    expect(mockModalRef.componentInstance.courseCandidates[0]).toEqual(testCourse2);
+    expect(mockModalRef.componentInstance.sessionToCopyCourseId).toEqual(testCourse1.courseId);
```

Comment:
What about testing this part?

```
modalRef.result.then((result: CopySessionModalResult) => {
        this.failedToCopySessions = {};
        const requestList: Observable<FeedbackSession>[] = this.createSessionCopyRequestsFromModal(
            result, this.courseId, this.feedbackSessionName);
        if (requestList.length === 1) {
          requestList[0].subscribe((createdSession: FeedbackSession) => {
            if (Object.keys(this.failedToCopySessions).length === 0) {
              this.navigationService.navigateWithSuccessMessage(this.router,
                  '/web/instructor/sessions/edit',
                  'The feedback session has been copied. Please modify settings/questions as necessary.',
                  { courseid: createdSession.courseId, fsname: createdSession.feedbackSessionName });
            } else {
              this.statusMessageService.showErrorToast(this.getCopyErrorMessage());
              this.sessionEditFormModel.isCopying = false;
            }
          }, (resp: ErrorMessageOutput) => { this.statusMessageService.showErrorToast(resp.error.message); });
        }
        if (requestList.length > 1) {
          forkJoin(requestList)
           .pipe(finalize(() => this.sessionEditFormModel.isCopying = false))
           .subscribe(() => {
             this.showCopyStatusMessage();
           });
        }
      }, (resp: ErrorMessageOutput) => { this.statusMessageService.showErrorToast(resp.error.message); })
      .catch(() => this.sessionEditFormModel.isCopying = false);
    });
```

Reply:
I have updated the test and tested this part by spying on navigation service